snapshots/windows/windows.go

```diff
@@ -403,29 +421,33 @@ func (s *snapshotter) createSnapshot(ctx context.Context, kind snapshots.Kind, k
 					return nil, fmt.Errorf("failed to parse label %q=%q: %w", rootfsSizeInBytesLabel, sizeBytesStr, err)
 				}
 			}
+		} else {
+			// A view is just a read-write snapshot with a _really_ small sandbox, since we cannot actually
+			// make a read-only mount or junction point. https://superuser.com/q/881544/112473
+			sizeInBytes = 20 * 1024 * 1024 * 1024
```

Comment:
I was thinking about ProjFS the other day, but since [AFAIK it expects you to copy the relevant file from the backing store into the directory it's mounted at to be used by the system](https://learn.microsoft.com/en-us/windows/win32/projfs/cache-state), it seems likely to be wasteful if I used it here when I already have the files available on the filesystem via the WCIFS driver.

I think a more-likely improvement would be being able to mount a WCIFS filesystem stack read-only, i.e. with no sandbox. I expect that would require OS-level support, unless there's something hidden in the HCS or [computestorage](https://learn.microsoft.com/en-us/virtualization/api/hcs/reference/hcsattachlayerstoragefilter) APIs that would allow this, and we just don't expose it in hcsshim.

Using ProjFS instead of WCIFS to actually mount a layer stack on the other hand would be a fascinating and terrifying exercise. 100% that would be an external snapshotter implementation that would need to be bug-for-bug compatible with WCIFS and anything else used to expose layers into the HCS computer systems (VM, container, or host-based). More like a research project than likely to become _the_ way, I think. And it would still have the problem that the current WCIFS on-disk format has all the files we need already extracted. It already has a triple-the-disk-space issue when exporting a layer (although I reckon once I have the Snapshotter test suite passing here, I can get that down to double-the-disk-space), and using ProjFS would introduce an up-to-double-the-disk-space issue in the mount-time behaviour.

Final thought, the liberal use of "legacy" in the related hcsshim code and discussions suggests that someone might have plans for a different layer-storage format, and that might be more-conducive to something like ProjFS for its host-mount system. But given that this stuff happens with Windows Server support lifecycle iterations, I don't expect any such work to materialise before I land this. (I also didn't expect to not have landed this last year, though, so take my expectations with a grain of salt. ^_^)

Reply:
Yeah, it wouldn't be the best option. I was thinking it might be of use for RO bind mounts, but that's not the only thing that is missing and would require RO access. WCIFS sounds like the path forward, but I am not sure if that filter driver has any flag that would allow setting a RO flag. I don't think it would be a huge technical challenge to add, but it might take a non-trivial amount of time until it propagates to all supported versions of Windows.

Reply:
Something else to look into is whether we can replace the symlink/junction-point appear with the [Bind Filter](https://github.com/microsoft/hcsshim/blob/2a2bfdd7a42b23dc0cb7e4ee7fe375346d70639d/internal/jobobject/jobobject.go#L471-L519).

I had seen that before, but only came across this API (and noticed it supports readonly binds) when just now preparing a hcsshim/WCIFS feature request.

We'd still like the WCIFS/hcsshim feature request, of course, since Bind Filter doesn't solve the problem of needing a scratch directory for `HcsAttachLayerStorageFilter` to initialise WCIFS. But it could be used to [optimise the "no parents" cases, as is done in the containerd overlay driver](https://github.com/containerd/containerd/blob/58701f87c7db5abf5c203a4a8d8b8558977d06cf/snapshots/overlay/overlay.go#L491-L509), at least.

Reply:
Which is already a huge plus. It would eliminate the need for a bunch of workarounds. Great find!

What are the blocking issues that remain to be solved to land both this and [the hcsshim PR](https://github.com/microsoft/hcsshim/pull/901) this depends on? If I can, I would like to help out.

Reply:
See https://github.com/microsoft/hcsshim/pull/901#issuecomment-1211456887 for what was left to do there. It's not specifically blocking issues in the code, as much as tests needing to be written and tested, combined with me currently being in the middle of a large move, and so likely to be non-contributing until mid-October _at least_.

I'd also like to not land that PR until this PR is passing the snapshotter test suite, since it works that code over _much_ more thoroughly than any tests I've written.

Somewhat parallelisable, the task-list on this PR includes a bunch of work to clean up some of the hacks/changes I've made to get this to mostly-pass the snapshotter test suite. The main block of work I was going to come to next was fixing continuity (or the containerd usage thereof) so that the volume mounts don't interfere with its file-walking, and hence revert the nastiest hack I added to the test suite, aka `wcow_workaround`.

Reply:
It seems that [continuity PR](https://github.com/containerd/continuity/pull/208) helps with comparing files, but now fails due to not being able to checksum hidden system files. See test output:

https://paste.samfira.com/public/p/dznNWyKUHCKT9s2rEJPKRhRK

Only ```SYSTEM``` has access to those files. Explicitly adding permissions for the user running the tests to access those files, should work. I'm wondering if we should create a list of files/folders that should be ecluded from comparison. ```$Recicle.Bin``` comes to mind, as well as things like ```pagefile.sys```. What are your thoughts on this?

Reply:
Setting proper dacl allows us to create a manifest of the entire mount point, and now the fun starts. We have some extra paths inside that mount point that makes the test fail:

```bash
PS C:\containerd> sh.exe -c "EXTRA_TESTFLAGS=-test.run=TestSnapshotterClient make integration"
+ integration
Windows test image: mcr.microsoft.com/windows/nanoserver:ltsc2022 , Windows build version: 20348
time="2022-09-27T07:45:07-07:00" level=info msg="running tests against containerd" revision=03536e7d953416c460ab076d497eadade3084ff1.m runtime= snapshotter= version=v1.6.0-715-g03536e7d9.m
time="2022-09-27T07:45:07-07:00" level=info msg="start to pull seed image" image="mcr.microsoft.com/windows/nanoserver:ltsc2022"
time="2022-09-27T07:45:08-07:00" level=info msg=request digest="sha256:09629875cd6ee57fa551626fa2963d8b54718ba9acbf28e7cb5d684a9cb754d7" mediatype=application/vnd.docker.image.rootfs.foreign.diff.tar.gzip size=118131331 url="https://mcr.microsoft.com/v2/windows/nanoserver/blobs/sha256:09629875cd6ee57fa551626fa2963d8b54718ba9acbf28e7cb5d684a9cb754d7"
=== RUN   TestSnapshotterClient
=== RUN   TestSnapshotterClient/Basic
    helpers.go:88: unmount C:\Users\ADMINI~1\AppData\Local\Temp\snapshot-suite-SnapshotterClient-2124266799\work\preparing
    helpers.go:88: unmount C:\Users\ADMINI~1\AppData\Local\Temp\snapshot-suite-SnapshotterClient-2124266799\work\nextlayer
    testsuite.go:216: failure reason: directory diff between C:\Users\ADMINI~1\AppData\Local\Temp\fstest2813155162 and C:\Users\ADMINI~1\AppData\Local\Temp\snapshot-suite-SnapshotterClient-2124266799\work\nextlayer
        + \WcSandboxState
        + \WcSandboxState\Hives
        + \WcSandboxState\Hives\DefaultUser_Delta
        + \WcSandboxState\Hives\Sam_Delta
        + \WcSandboxState\Hives\Security_Delta
        + \WcSandboxState\Hives\Software_Delta
        + \WcSandboxState\Hives\System_Delta
        + \WcSandboxState\initialized
        + \Windows
        + \Windows\System32
        + \Windows\System32\config
        + \Windows\System32\config\DEFAULT
        + \Windows\System32\config\SAM
        + \Windows\System32\config\SECURITY
        + \Windows\System32\config\SOFTWARE
        + \Windows\System32\config\SYSTEM
    helpers.go:69: drwxrwxrwx          0 C:\Users\ADMINI~1\AppData\Local\Temp\snapshot-suite-SnapshotterClient-2124266799
    helpers.go:69: drwxrwxrwx          0 C:\Users\ADMINI~1\AppData\Local\Temp\snapshot-suite-SnapshotterClient-2124266799\root
    helpers.go:69: drwxrwxrwx          0 C:\Users\ADMINI~1\AppData\Local\Temp\snapshot-suite-SnapshotterClient-2124266799\work
    helpers.go:69: drwxrwxrwx          0 C:\Users\ADMINI~1\AppData\Local\Temp\snapshot-suite-SnapshotterClient-2124266799\work\nextlayer
```

Your current workaround may be needed after all, but this time for different reasons :smile: .

Reply:
I'm surprised `WcSandboxState` exists at all, I'd have assumed that only existed when the layers are used by a running container. I wonder, are those files all empty? Since no OS has been booted in that image, I can't see how there'd be any "delta" to capture, but maybe a zero delta has non-zero size.

It would make sense to me that for purposes of the snapshotter diff test,  `\WcSandboxState` and `\Windows` be ignored; I'd say specifically `\Windows\System32\config\XXX` for the five registry hives (which I expect are trivially bind-mounted from the directory next to Files in the underlying layer image) but then having to ignore the containing directories sounds messy.

An alternative would be to do something clever with the compared-against version, using an applier or something to generate empty Deltas and `initialized` in `WcSandboxState`, and copy in the (empty in this case) hives from the underlying layer into `Windows\System32\config\`. That would mean we know exactly how those directories behave (since we can reproduce the behaviour and get a clean diff) but might tie the tests to a behaviour that varies across OS versions, and also means effectively reverse-engineering that behaviour. I don't think it's worth the effort to take this path (invoking [Hyrum's Law](https://www.hyrumslaw.com/) in the process), compared to excluding a fairly-fixed list of directories.